vpr/src/noc/sat_routing.cpp

```diff
@@ -0,0 +1,986 @@
+        src_map[src_router_id] = l;
+        is_dst[dst_router_id] = true;
+    }
+
+    // Find the starting pair (whose first element is not a second element of any pair)
+    auto it = links.begin();
+    for (; it != links.end(); ++it) {
+        NocRouterId src_router_id = noc_model.get_single_noc_link(*it).get_source_router();
+        if (is_dst.find(src_router_id) == is_dst.end()) {
+            break;
+        }
+    }
+
+    // Reconstruct the chain starting from the found starting pair
+    auto current = *it;
+    while (true) {
+        route.push_back(current);
+        NocRouterId dst_router_id = noc_model.get_single_noc_link(current).get_sink_router();
+        auto nextIt = src_map.find(dst_router_id);
+        if (nextIt == src_map.end()) {
+            break; // End of chain
+        }
+        current = nextIt->second;
+    }
+
+    VTR_ASSERT(route.size() == links.size());
+
+    return route;
+}
+
+static vtr::vector<NocTrafficFlowId, std::vector<NocLinkId>> convert_vars_to_routes(t_flow_link_var_map& flow_link_vars,
+                                                                                    const orsat::CpSolverResponse& response) {
+    const auto& noc_ctx = g_vpr_ctx.noc();
+    const auto& traffic_flow_storage = noc_ctx.noc_traffic_flows_storage;
+
+    VTR_ASSERT(response.status() == orsat::CpSolverStatus::FEASIBLE ||
+               response.status() == orsat::CpSolverStatus::OPTIMAL);
+
+    vtr::vector<NocTrafficFlowId, std::vector<NocLinkId>> routes;
+    routes.resize(traffic_flow_storage.get_number_of_traffic_flows());
+
+    for (auto& [key, var] : flow_link_vars) {
+        auto [traffic_flow_id, noc_link_id] = key;
+        bool value = orsat::SolutionBooleanValue(response, var);
+        if (value) {
+            routes[traffic_flow_id].push_back(noc_link_id);
+        }
+    }
+
+    for (auto& route : routes) {
+        route = sort_noc_links_in_chain_order(route);
+    }
+
+    return routes;
+}
+
+static void create_flow_link_vars(orsat::CpModelBuilder& cp_model,
+                                  t_flow_link_var_map& flow_link_vars,
+                                  std::map<NocTrafficFlowId , orsat::IntVar>& latency_overrun_vars) {
+    const auto& noc_ctx = g_vpr_ctx.noc();
+    const auto& noc_model = noc_ctx.noc_model;
+    const auto& traffic_flow_storage = noc_ctx.noc_traffic_flows_storage;
+
+    // TODO: specify the domain based on NoC topology
+    operations_research::Domain latency_overrun_domain(0, 20);
+
+    // create boolean variables for each traffic flow and link pair
+    // create integer variables for traffic flows with constrained latency
+    for (auto traffic_flow_id : traffic_flow_storage.get_all_traffic_flow_id()) {
+        const auto& traffic_flow = traffic_flow_storage.get_single_noc_traffic_flow(traffic_flow_id);
+
+        // create an integer variable for each latency-constrained traffic flow
```

Comment:
Comment why you're comparing to 0.1. Suggest constexpr UNREASONABLY_LARGE_LATENCY 0.1